opm/simulators/linalg/cuistl/CuSparseMatrix.cpp

```diff
@@ -0,0 +1,327 @@
+    T beta = 0.0;
+    OPM_CUSPARSE_SAFE_CALL(detail::cusparseBsrmv(m_cusparseHandle.get(),
+                                                 detail::CUSPARSE_MATRIX_ORDER,
+                                                 CUSPARSE_OPERATION_NON_TRANSPOSE,
+                                                 m_numberOfRows,
+                                                 m_numberOfRows,
+                                                 m_numberOfNonzeroBlocks,
+                                                 &alpha,
+                                                 m_matrixDescription->get(),
+                                                 nonzeroValues,
+                                                 rowIndices,
+                                                 columnIndices,
+                                                 blockSize(),
+                                                 x.data(),
+                                                 &beta,
+                                                 y.data()));
+}
+
+template <typename T>
+void
+CuSparseMatrix<T>::umv(const CuVector<T>& x, CuVector<T>& y) const
+{
+    assertSameSize(x);
+    assertSameSize(y);
+    if (blockSize() < 2u) {
+        OPM_THROW(
+            std::invalid_argument,
+            "CuSparseMatrix<T>::usmv and CuSparseMatrix<T>::mv are only implemented for block sizes greater than 1.");
+    }
+
+    const auto nonzeroValues = getNonZeroValues().data();
+
+    auto rowIndices = getRowIndices().data();
+    auto columnIndices = getColumnIndices().data();
+    T alpha = 1.0;
+    T beta = 1.0;
+    OPM_CUSPARSE_SAFE_CALL(detail::cusparseBsrmv(m_cusparseHandle.get(),
+                                                 detail::CUSPARSE_MATRIX_ORDER,
+                                                 CUSPARSE_OPERATION_NON_TRANSPOSE,
+                                                 m_numberOfRows,
+                                                 m_numberOfRows,
+                                                 m_numberOfNonzeroBlocks,
+                                                 &alpha,
+                                                 m_matrixDescription->get(),
+                                                 nonzeroValues,
+                                                 rowIndices,
+                                                 columnIndices,
+                                                 m_blockSize,
+                                                 x.data(),
+                                                 &beta,
+                                                 y.data()));
+}
+
+template <typename T>
+void
+CuSparseMatrix<T>::usmv(T alpha, const CuVector<T>& x, CuVector<T>& y) const
+{
+    assertSameSize(x);
+    assertSameSize(y);
+    if (blockSize() < 2) {
+        OPM_THROW(
+            std::invalid_argument,
+            "CuSparseMatrix<T>::usmv and CuSparseMatrix<T>::mv are only implemented for block sizes greater than 1.");
```

Comment:
This may be needed for CPR for example, where we have a pressure matrix with 1x1 blocks, or scalars. I assume it is not a fundamental design and can be fixed when needed.

Reply:
It would mean changing to a CSR and not BSR (block CSR) matrix. Can be done with some rewriting.